sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/StartingPointShardsFinder.java

```diff
@@ -0,0 +1,202 @@
+   *   its successor. 0007 is valid at T3 timestamp and it's added to the result set.</li>
+   *   <li>0002 is closed at T3 timestamp so its successors 0005 and 0006 are next. 0005 is also
+   *   closed because it was merged with 0004 shard. Their successor is 0007 and it was already
+   *   considered by 0004 shard so no action here is needed. Shard 0006 is valid.</li>
+   * </ul>
+   * </li>
+   * <li>T4 timestamp (or LATEST marker) - following the same reasoning as in previous cases it
+   * end's up with 0000, 0003, 0008 and 0010 shards.</li>
+   * </ul></p>
+   *
+   * <pre>
+   *      T1                T2          T3                      T4
+   *      |                 |           |                       |
+   * 0000-----------------------------------------------------------
+   *
+   *
+   *             0003-----------------------------------------------
+   *            /
+   * 0001------+
+   *            \
+   *             0004-----------+             0008------------------
+   *                             \           /
+   *                              0007------+
+   *                             /           \
+   *                  0005------+             0009------+
+   *                 /                                   \
+   * 0002-----------+                                     0010------
+   *                 \                                   /
+   *                  0006------------------------------+
+   * </pre>
+   */
+  Iterable<Shard> findShardsAtStartingPoint(SimplifiedKinesisClient kinesis, String streamName,
+      StartingPoint startingPoint) throws TransientKinesisException {
+    List<Shard> allShards = kinesis.listShards(streamName);
+    Set<Shard> initialShards = findInitialShardsWithoutParents(streamName, allShards);
+
+    Set<Shard> startingPointShards = new HashSet<>();
+    Sets.SetView<Shard> expiredShards;
+    do {
+      Set<Shard> validShards = validateShards(kinesis, initialShards, streamName, startingPoint);
+      startingPointShards.addAll(validShards);
+      expiredShards = Sets.difference(initialShards, validShards);
+      Set<String> expiredShardIds = FluentIterable.from(expiredShards)
+          .transform(new ShardIdGetter())
+          .toSet();
+      if (!expiredShardIds.isEmpty()) {
+        LOGGER.info("Following shards expired for {} stream at '{}' starting point: {}", streamName,
+            startingPoint, expiredShardIds);
+      }
+      initialShards = FluentIterable.from(allShards)
+          .filter(new ExpiredShardsSuccessorsFilter(expiredShardIds))
+          .toSet();
+    } while (!expiredShards.isEmpty());
+    return startingPointShards;
+  }
+
+  /**
+   * Finds the initial set of shards (the oldest ones). These shards do not have their parents in
+   * the shard list.
+   */
+  private Set<Shard> findInitialShardsWithoutParents(String streamName, List<Shard> allShards) {
+    final Set<String> shardIds = FluentIterable.from(allShards)
```

Comment:
Once again seems like a matter of personal style, but - even speaking as a huge fan of FP - to me this method would look better as a simple single Java loop without usage of FluentIterable. Currently we're wasting dozens of lines on creating trivial classes for wrapping 1-line logic and scattering the logic over more than 1 screen where it could have fit in a few lines.

Reply:
Done, I admit it looks better this way